app/health_check.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+)
+
+type PluginHealthCheckJob struct {
+	server    *Server
+	task      *model.ScheduledTask
+	taskMutex sync.Mutex
+}
+
+const (
+	PLUGIN_HEALTH_CHECK_TASK_NAME = "Plugin Health Check"
+	PLUGIN_HEALTH_CHECK_INTERVAL  = 5 // seconds
+	HEALTH_CHECK_FAIL_LIMIT       = 3
+	ENABLE_HEALTH_CHECK_JOB       = false
+)
+
+func (s *Server) InitPluginHealthCheckJob() {
+	if ENABLE_HEALTH_CHECK_JOB {
+		s.PluginsLock.RLock()
+		pluginsEnvironment := s.PluginsEnvironment
+		s.PluginsLock.RUnlock()
+
+		if pluginsEnvironment == nil {
+			return
+		}
+
+		if s.PluginHealthCheck == nil {
+			s.PluginHealthCheck = NewPluginHealthCheckJob(s)
+		}
+
+		s.PluginHealthCheck.Start()
+	}
+}
+
+func (job *PluginHealthCheckJob) Start() {
+	mlog.Debug(fmt.Sprintf("Plugin health check job starting. Sending health check pings every %v seconds.", PLUGIN_HEALTH_CHECK_INTERVAL))
+	newTask := model.CreateRecurringTask(PLUGIN_HEALTH_CHECK_TASK_NAME, job.RunPluginHealthChecks, time.Duration(PLUGIN_HEALTH_CHECK_INTERVAL)*time.Second)
```

Comment:
Not sure the job system is the right fit here. It's designed to run once on an entire cluster and possibly on a separate server. But this needs to run on each server.

Reply:
That makes sense, thanks for pointing this out. Do you have ideas on how to schedule the health check per server?

Reply:
I don't think we have a specific pattern for that. Just a regular go routine would work.

Reply:
@levb I agree. I'm thinking of how to solve this. It would likely be a function of the health check interval and the number of plugins. Or possibly an explicitly configured value for the time distance between each plugin's check. There is also the situation where we check upon visiting the admin console. In this case, they would be simulataneous yet concurrent. In the health check job situation, they would also be concurrent, if simulataneous.